pkg/core/native/native_neo.go

```diff
@@ -828,6 +835,34 @@ func (n *NEO) registerCandidate(ic *interop.Context, args []stackitem.Item) stac
 	return stackitem.NewBool(err == nil)
 }
 
+func (n *NEO) onNEP17Payment(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	var (
+		caller = ic.VM.GetCallingScriptHash()
+		_      = toUint160(args[0])
+		amount = toBigInt(args[1])
+		pub    = toPublicKey(args[2])
+	)
+
+	if caller != n.GAS.Hash {
+		panic("only GAS is accepted")
+	}
+	if !amount.IsInt64() || amount.Int64() != n.getRegisterPriceInternal(ic.DAO) {
+		panic(fmt.Errorf("incorrect GAS amount for registration (expected %d)", n.getRegisterPriceInternal(ic.DAO)))
+	}
+	ok, err := runtime.CheckKeyedWitness(ic, pub)
+	if err != nil {
+		panic(err)
+	} else if !ok {
+		panic("not witnessed by the key owner")
+	}
+	n.GAS.burn(ic, n.Hash, amount)
+	err = n.RegisterCandidateInternal(ic, pub)
```

Comment:
I have concern about the order of these methods. No reentry is possible here since `n.GAS.burn` doesn't call `onNEP17Payment`, but following the logic of native NEP17, have you consider firstly to update the candidate's state, and only after that perform GAS burn? The order of notifications then will be different: [CandidateStateChanged, Transfer].

But it's probably not the most important thing.

Reply:
The only thing changed is event order then, I'm pretty agnostic to any particular choice.

Reply:
I still think that other solution should be possible

Reply:
For the testing seems a big change to me, but for the other issue, I agree that is hard to solve different than this